### PR TITLE
[API-116] Tenant support in Unity

### PR DESCRIPTION
**Pull request by Tianqi-Tang**
Add tenant and version support into storops.

### CODE REVIEW

File: storops/lib/version.py [hunk 1]
```diff
@@ -0,0 +1,305 @@
...
+    >>          return '<=2.0'
+
+    :param criteria_string: The version criteria expression. The supported
+         comparator includes: <, <=, =, ==, !=, >, >=, <>(between)
+    """
+    return Version(criteria_string, get_version)
+
+
+class Version(object):
+    INIT_METHOD = 1
```
Comment (jealous) [inline storops/lib/version.py]: It looks like an Enum.  Could we make it an Enum.

File: storops/lib/version.py [hunk 2]
```diff
@@ -0,0 +1,305 @@
...
+        return self._handle_method(Version.INIT_METHOD, func)
+
+    def _handle_method(self, dec_type, func):
+        def _inner(*args, **kwargs):
+            self._test_support(dec_type, *args, **kwargs)
+            return func(*args, **kwargs)
+
+        return _inner
+
+    def _test_support(self, *args, **kwargs):
```
Comment (jealous) [inline storops/lib/version.py]: Where did you use `dec_type`?

File: test/lib/test_version.py [hunk 3]
```diff
@@ -0,0 +1,218 @@
...
+    def test_new_instance_unsupport(self):
+        new_obj = functools.partial(DemoResource1, t_rest('2'))
+        assert_that(new_obj, raises(SystemAPINotSupported))
+
+    def _test_cls_or_static_method_support(self, cli, method):
+        assert_that(getattr(DemoResource1, method)(cli, 'a'), 'a')
+
+    def _test_cls_or_static_method_unsupport(self, cli, method):
+        call = functools.partial(getattr(DemoResource1, method), cli, 'a')
+        assert_that(call, SystemAPINotSupported)
```
Comment (jealous) [inline test/lib/test_version.py]: missing `raises`.

File: storops/lib/version.py (hunk 2 above)
Comment (Tianqi-Tang) [inline storops/lib/version.py]: The dec_type will be passed to the version getter as the first parameter. The version getter could use it to determine how to get the version.

File: test/lib/test_version.py (hunk 3 above)
Comment (Tianqi-Tang) [inline test/lib/test_version.py]: Done, thanks.

File: storops/lib/version.py (hunk 1 above)
Comment (Tianqi-Tang) [inline storops/lib/version.py]: They are flags for version getters to know the type of the decorator.  Per the discussion offline, I will leave it as it is.

File: storops/lib/version.py [hunk 4]
```diff
@@ -0,0 +1,305 @@
...
+        if (isinstance(obj, Resource) and
+                (dec_type == Version.INSTANCE_METHOD or
+                 dec_type == Version.METHOD)):
+            return obj.system_version
+    return _get_version_from_client_in_args(*args, **kwargs)
+
+
+def _get_version_from_client_in_args(*args, **kwargs):
+    for parm in list(args) + list(kwargs.values()):
+        if isinstance(parm, (UnityClient, CliClient)):
```
Comment (jealous) [inline storops/lib/version.py]: You could replace this line with `hasattr(parm, 'system_version')`.
And then we don't need to import `UnityClient` or `CliClient`.

File: storops/lib/version.py [hunk 5]
```diff
@@ -0,0 +1,305 @@
...
+    def _test_support(self, dec_type, *args, **kwargs):
+        version = self.version_getter(dec_type, *args, **kwargs)
+        if not self.criteria.test(version):
+            raise SystemAPINotSupported(version=version)
+
+
+def get_version(dec_type, *args, **kwargs):
+    if len(args) > 0:
+        obj = args[0]
+        if (isinstance(obj, Resource) and
```
Comment (jealous) [inline storops/lib/version.py]: Same here, we could just check whether `obj` has an attribute named `system_version`.  This is the only dependency we have, we don't care if it's an instance of certain class.

File: storops/lib/version.py (hunk 4 above)
Comment (Tianqi-Tang) [inline storops/lib/version.py]: It doesn't work because the Classes also has the attribute system_version

File: storops/lib/version.py (hunk 5 above)
Comment (Tianqi-Tang) [inline storops/lib/version.py]: It doesn't work because the Classes also has the attribute system_version

Comment (Tianqi-Tang) [inline storops/lib/version.py]: It doesn't work because the class will also has the system_version property.

File: storops/lib/version.py (hunk 4 above)
Comment (Tianqi-Tang) [inline storops/lib/version.py]: It doesn't work because the class will also has the system_version property.

**Pull request merged**